### PR TITLE
Feature/custom contract data

### LINKED ISSUE #214: [TX-Builder] Allow custom contract data
## Overview
Currently in the Transaction Builder we only allow to create new transactions by selecting and filling the complete information after an ABI is provided and all fields are filled. Some users could prepare some custom data in some other dapp and just be willing to send it from the TX-Builder, for example sending a transaction where the signatures were gathered off chain

## Goals
We should allow switching to show custom hex encoded data the same way we allow it in contract interaction, either converting the user input to hex data or allowing to send custom hex data

## Requirements
- [ ] Allow input custom hex data to add a new transaction to the transaction builder
- [ ] Show custom hex data when user filled the different inputs when interacting with a smart contract

## Screens
Example in "Contract interaction" used as TX-Builder
image

Example in "Contract interaction" of custom hex data transaction
imagen

Any doubts we should check contract interaction feature on Safe web

### PR DESCRIPTION
**Pull request by yagopv**
## What it solves
Resolves #214 

## How this PR fixes it
We are adding a Switch for allowing the user to introduce custom data

When the user activates the switch then a `textarea` is presented. If the user was previously selected a method and filled the individual inputs and the validation result is positive then we are prefilling the textarea with the custom data.

Adding as well a "Show more/less" link in the preview modal before sending the transactions to the safe-web

We can send some Native token amount with the custom transaction data

## How to test it
1) Open TX Builder
2) <ins>Case 1</ins>. Select a method, fill the inputs with correct data and before adding the transaction toggle the switch => The custom data should be prefilled. Test the reverse case where the user is prefilling something wrong and then the textarea is not prefilled. Add the transaction
3) <ins>Case 2</ins>. Toggle the switch and enter the custom data. Should be hex strict or an error will be showed. Add transaction
4) <ins>Case 3</ins>. Add custom data with a length greater than 50 characters. In the popup with the transactions resume, the custom data with a "Show more/less" link should be showed
5) <ins>Case 4</ins>. Send some native token amount with the custom data and check is reflected in the final safe-web transaction

## Screenshots

image

image

### CODE REVIEW

File: apps/tx-builder/src/components/Builder.tsx [hunk 1]
```diff
@@ -263,34 +289,46 @@ export const Builder = ({
...
   }, [transactions]);
 
+  useEffect(() => {
+    if (showCustomData) {
+      setCustomDataValue('');
+      try {
+        const txData = getTxData();
+
+        if (txData) {
+          if (services.web3 && !services.web3.utils.isHexStrict(txData.data as string)) {
```
Comment (mikheevm) [inline apps/tx-builder/src/components/Builder.tsx:313]: you could use typescript's optional chaining here (and a few other places). Just a small nitpick.

File: apps/tx-builder/src/components/Builder.tsx [hunk 2]
```diff
@@ -48,6 +64,10 @@ const StyledAddressInput = styled(AddressInput)`
   }
 `;
 
+const StyledTextAreaField = (props: any) => {
```
[the remark this answers is not in the document]
Comment (yagopv) [reply apps/tx-builder/src/components/Builder.tsx:67]: super true ☺️

File: apps/tx-builder/src/components/Builder.tsx [hunk 3]
```diff
@@ -170,12 +144,29 @@ export const Builder = ({
     setReviewing(false);
   };
 
-  const addTransaction = async () => {
+  const handleCustomDataInputChange = useCallback(
+    (e: any) => {
```
[the remark this answers is not in the document]
Comment (yagopv) [reply apps/tx-builder/src/components/Builder.tsx:148]: right!

File: apps/tx-builder/src/utils.ts [hunk 4]
```diff
@@ -29,3 +31,56 @@ export const rpcUrlGetterByNetwork: {
   [CHAINS.ARBITRUM]: () => 'https://arb1.arbitrum.io/rpc',
   [CHAINS.VOLTA]: () => 'https://volta-rpc.energyweb.org',
 };
+
+export const getInputHelper = (input: any) => {
```
[the remark this answers is not in the document]
Comment (yagopv) [reply apps/tx-builder/src/utils.ts:35]: i'll do it no problem!

**Review by mikheevm — APPROVED** [review]

**Comment by JagoFigueroa** [comment]
Quick note from my end. I don't think we should show the 'use custom data' when the address does not have an ABI what do you think @yagopv @dasanra ? 

Screenshot 2021-12-13 at 11 19 45

**Comment by mikheevm** [comment]
> Quick note from my end. I don't think we should show the 'use custom data' when the address does not have an ABI what do you think @yagopv @dasanra ?
> 
> Screenshot 2021-12-13 at 11 19 45

This is exactly the situation where you need the custom data :)

**Comment by JagoFigueroa** [comment]
All good from my end ;)

**Pull request merged**